### PR TITLE
PMM-9200: Test SMTP configuration

### PR DESCRIPTION
**Pull request by fabio-silva**
**What this PR does / why we need it**: Tests a SMTP configuration before setting it up.

**Which issue(s) this PR fixes**: https://jira.percona.com/browse/PMM-9200

Approach:
- We needed an input field "within" the form (test email) to have access to its values, but that field itself should not be part of the form values and shouldn't make the form invalid in any sense;
- Although a form inside a form would work, it's not a recommended strategy;
- Using a portal, that input can be rendered outside the form, while still having the context (values) from the form;
- By doing this, the "test email" part would be rendered after the form (i.e. after the "Apply" button"), which is not very pleasant, from the UI perspective;
- Another portal was created to render the "Apply" button after the test input. By doing this, the input is rendered before the submit button, while both have access to the form's context.

Screenshot 2021-11-25 at 11 52 33

### CODE REVIEW

File: public/app/percona/settings/Settings.tsx [hunk 1]
```diff
@@ -30,6 +30,8 @@ export const SettingsPanel: FC = () => {
     refresh?: boolean,
     onError = () => {}
   ) => {
+    const testEmail = body.email_alerting_settings.test_email || '';
+    body.email_alerting_settings.test_email = undefined;
```
Comment (tiagomotasantos) [inline public/app/percona/settings/Settings.tsx:34]: Why do we need to remove the test_email from the body?

File: public/app/percona/settings/components/Communication/Email/Email.tsx [hunk 2]
```diff
@@ -128,18 +146,35 @@ export const Email: FC<EmailProps> = ({ updateSettings, settings }) => {
...
+                >
+                  {loading && <Spinner />}
+                  {Messages.actionButton}
+                </Button>,
+                applyRef.current
+              )}
           </form>
         )}
       />
+      <div ref={(e) => (testRef.current = e)}></div>
```
Comment (tiagomotasantos) [inline public/app/percona/settings/components/Communication/Email/Email.tsx:176]:
```suggestion
      <div ref={testRef}></div>
```

Isn't this the same?

File: public/app/percona/settings/Settings.tsx (hunk 1 above)
Comment (fabio-silva) [reply public/app/percona/settings/Settings.tsx:34]: Yeah, this one is hard to follow, I'll add a comment there.
The thing is: when you submit the form, you don't want the values there to vanish after the re-render. All other fields stay there because we use the values coming from the API response. But for the test email that wouldn't happen, as that field is not part of the request. So, the value is being sent all the way up to the parent component (`Settings.tsx`) so that it can be later on sent down with the payload again. But since we don't want to send this field in the request payload, I'm just erasing it here.

File: public/app/percona/settings/components/Communication/Email/Email.tsx (hunk 2 above)
Comment (fabio-silva) [reply public/app/percona/settings/components/Communication/Email/Email.tsx:176]: Well, the problem is that, initially, the ref will be null. We could either use `useState` to observe changes on `testRef` and make the component re-render, or just simply use a callback ref: https://reactjs.org/docs/refs-and-the-dom.html#callback-refs
Doing this, by the time the component mounts, it's guaranteed our ref is updated, as stated in the docs.

File: public/app/percona/settings/components/Communication/Email/TestEmailSettings/TestEmailSettings.styles.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  form: css`
+    display: flex;
+  `,
+  input: css`
+    flex-grow: 1;
+  `,
+  button: css`
+    height: 37px;
+    margin-left: ${spacing.md};
+    margin-top: 20px;
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/percona/settings/components/Communication/Email/TestEmailSettings/TestEmailSettings.styles.ts:14]: Hm, not in this case. I'm using this margin to make sure the button is aligned with the input on the left. It's a bit hacky, but it's the only thing we can do until we have some sort of our own component.

File: public/app/percona/settings/components/Communication/Email/TestEmailSettings/TestEmailSettings.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  const handleClick = async (email: string) => {
+    setTestingSettings(true);
+    await onTest(email);
+    setTestingSettings(false);
+  };
+
+  return (
+    <Form
+      onSubmit={() => {}}
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/percona/settings/components/Communication/Email/TestEmailSettings/TestEmailSettings.tsx:21]: `TextInputField` has to be inside a `Form` for it to work. But we don't actually don't need a form, we just want the input and the button. But `Form` needs a `onSubmit` prop, so I just passed an empty callback. The idea here is not to actually trigger a form submission, but just hold the value from the input and make an API call

File: public/app/percona/settings/components/Communication/Communication.service.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { EmailPayload } from '../../Settings.types';
+import { api } from 'app/percona/shared/helpers/api';
+
+export const CommunicationService = {
+  async testEmailSettings(settings: EmailPayload, email: string): Promise<void> {
+    return api.post('/v1/Settings/TestEmailAlertingSettings', { ...settings, email_to: email });
```
Comment (solovevayaroslavna) [inline public/app/percona/settings/components/Communication/Communication.service.ts:6]: Maybe I'm wrong with the repository... In grafana, we don't have an agreement to put all the endpoints in one place?

Comment (fabio-silva) [reply public/app/percona/settings/components/Communication/Communication.service.ts:6]: well, not yet. We do have that in saas-ui, but not in grafana. But it's an idea for a future improvement :))

**Review by solovevayaroslavna — APPROVED** [review]

File: public/app/percona/settings/Settings.tsx (hunk 1 above)
Comment (tiagomotasantos) [reply public/app/percona/settings/Settings.tsx]: Thank you for the explanation!

**Review by tiagomotasantos — APPROVED** [review]
